specs/logging_implementation.adoc

```diff
@@ -0,0 +1,174 @@
+* A schema needs to be defined for emitted messages. This could be a json schema.
+  The schema needs to include metadata that allows the exact point of origin of
+  the message to be known.
+* Message priorities, types etc. need to be defined at the time of emission.
+* The integration/bridge applications must parse and restructure messages emitted
+  by the storage system to fit in the tendrl message schema before emitting it.
+* For messages that serve as notifications to the administrator (or alerts),
+  exact error message must be hardcoded by the developers. No backtraces or
+  system generated error messages should be sent as notifications to the
+  administrator.
+* Notifications sent to the administrator should provide sufficient context to
+  enable debugging of the problem via investigation into the log files.
+* Notifications and alerts need to be delivered to the node agent alert sockets.
+* Commons needs to implement a logging library that hides all the implementation
+  details and configuration options from any of the components using the library.
+  Components should be able to blindly emit messages to the common logging module
+  and it should be able to deliver them to either syslog, log files, alert
+  sockets etc. depending upon the metadata and configuration.
+* Syslog must be the default logging configuration. Logging to individual
+  configuration files should be thoroughly discouraged.
+
+== Use Cases
+
+Logs and alerts need to be treated as messages.
+
+== Proposed change
+* Create a new class called "Message" in common with attributes name, type,
+  message, priority(empty for log), alert_id ,log_level(empty for alert),
+  function_name, line_number.
+  (Logging will take line number and function name from where it is called, Here
+  the problem is we are logging the message from common so it takes a function name
+  and line number from common only. So it makes debugging complex. To avoid this
+  we have to find the function name, line number, file name and append this data
+  with message manually.
+```
+from inspect import getframeinfo, stack
+caller = getframeinfo(stack()[1][0])
+caller.filename, caller.lineno, caller.function
+(These are built in packages in python)
+```
+* Using name retrieve the particular logging handler.
+* Create a new function for Message class and process the object values.
+```
+1. If the type is Alert then create a new uuid and assign the uuid to alert. After
+   that convert the object into json, If the json is created successfully then store
+   json into log file as warning and put the json in node_agent socket also.
+2. If the type is Log then process the object and store json in log file based on
+   log level.
+```
+* Import the Message class in all the modules and Change log statements to
+  populate the instance of message class and pass the object to function.
+* Expected output format for log and alert in log file:
```

Comment:
what about sending logs to syslog? What will the logs/alerts look like in syslog/journal? Will you be using any special systemd journal fields?